reconcile/templating/renderer.py

```diff
@@ -0,0 +1,172 @@
+        query_func = gql.get_api().query
+    return query(query_func).template_collection_v1 or []
+
+
+class TemplateOutput(BaseModel):
+    path: str
+    content: str
+
+
+class FilePersistence(ABC):
+    @abstractmethod
+    def write(self, outputs: list[TemplateOutput]) -> None:
+        pass
+
+    @abstractmethod
+    def read(self, path: str) -> str:
+        pass
+
+
+class LocalFilePersistence(FilePersistence):
+    def __init__(self, app_interface_data_path: str) -> None:
+        self.app_interface_data_path = app_interface_data_path
+
+    def write(self, outputs: list[TemplateOutput]) -> None:
+        for output in outputs:
+            with open(
+                f"{join_path(self.app_interface_data_path, output.path)}",
+                "w",
+                encoding="utf-8",
+            ) as f:
+                f.write(output.content)
+
+    def read(self, path: str) -> str:
+        with open(
+            f"{join_path(self.app_interface_data_path, path)}", "r", encoding="utf-8"
+        ) as f:
+            return f.read()
+
+
+def unpack_static_variables(
+    collection_variables: TemplateCollectionVariablesV1,
+) -> dict:
+    variables = {}
+    if collection_variables.static:
+        variables = collection_variables.static
+    return variables
+
+
+def unpack_dynamic_variables(
+    collection_variables: TemplateCollectionVariablesV1,
+) -> dict:
+    variables: dict[str, Any] = {}
+    if collection_variables.dynamic:
+        gql_no_validation = init_from_config(validate_schemas=False)
+        for dv in collection_variables.dynamic or []:
+            variables[dv.name] = []
+            result = gql_no_validation.query(dv.query)
+            for key in result or {}:
+                variables[dv.name].append(result[key])
+    return variables
+
+
+class TemplateRendererIntegrationParams(PydanticRunParams):
+    app_interface_data_path: Optional[str]
+
+
+def join_path(base: str, sub: str) -> str:
+    # not using os.path.sep, since app-interface relies on unix paths
+    if sub.startswith(APP_INTERFACE_PATH_SEPERATOR):
+        return os.path.join(base, sub[1:])
+    return os.path.join(base, sub)
+
+
+class TemplateRendererIntegration(QontractReconcileIntegration):
+    def __init__(self, params: TemplateRendererIntegrationParams) -> None:
+        super().__init__(params)
+
+    @staticmethod
+    def process_template(
+        template: TemplateV1,
+        variables: dict,
+        persistence: FilePersistence,
+    ) -> Optional[TemplateOutput]:
+        r = create_renderer(
+            template,
+            TemplateData(
+                variables=variables,
+            ),
+        )
+        target_path = r.render_target_path()
+        current_str: Optional[str] = None
+        try:
+            current_str = persistence.read(
+                target_path,
+            )
+            y = yaml.YAML()
+
+            r.data.current = y.load(current_str)
+        except FileNotFoundError:
```

Comment:
`FileNotFoundError` is only for `persistence.read`, right? Should limit try block only include expected exceptions.

Then no need to specify `current_str` outside, as it will early raise, won't be None when used later.

Reply:
Right, makes sense especially, since persistence would also be remote gitlab in the future. I'll change `read` to return `Optional[str]` and catch that exception.